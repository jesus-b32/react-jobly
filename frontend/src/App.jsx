import { useState, useEffect } from 'react';
// import './App.css';
import JoblyApi from './api';
import NavBar from './navbar/NavBar';
import ReactRoutes from './routes/ReactRoutes';
import { jwtDecode } from 'jwt-decode';
import UserContext from './auth/UserContext';
import useLocalStorage from './hooks/useLocalStorage';

function App() {
  const [isLoading, setIsLoading] = useState(true);
  //store user details of current user logged in; updated when user submits login form or signup form
  const [currentUser, setCurrentUser] = useState(null);
  //store user details of current user logged in; updated when user submits login form or signup form
  const [token, setToken] = useLocalStorage('currUserToken');
  //store the job ids that user as applied for; updated when user clicks on apply button in jobs page or company detail page
  const [jobIds, setJobIds] = useState(new Set([]));

  /**
   * Load user info from API. Until a user is logged in and they have a token,
  this should not run. It only needs to re-run when a user logs out, so
  the value of the token is a dependency for this effect.
   */
  useEffect(() => {
    async function getCurrentUser() {
      if (token) {
        try {
          const { username } = jwtDecode(token);
          // put the token on the Api class so it can use it to call the API.
          JoblyApi.token = token;
          const currentUser = await JoblyApi.getUser(username);
          setCurrentUser(currentUser);
          setJobIds(new Set(currentUser.applications));
        } catch (err) {
          console.error('App load User Info: problem loading', err);
          setCurrentUser(null);
        }
      }
      setIsLoading(false);
    }
    /** set isLoading to false while getting user details;
     * once the data is fetched (or even if an error happens!), this will be set backto false to dsiable loading screen.*/
    setIsLoading(true);
    getCurrentUser();
  }, [token]);

  /** Handles signup.
   * Automatically logs user in (set token) upon signup.
   */
  async function signup(signupData) {
    try {
      const token = await JoblyApi.signup(signupData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error('signup failed', errors);
      return { success: false, errors };
    }
  }

  /** Handles login.*/
  async function login(loginData) {
    try {
      const token = await JoblyApi.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error('login failed', errors);
      return { success: false, errors };
    }
  }

  /** Handles logout. */
  function logout() {
    setCurrentUser(null);
    setToken(null);
  }

  /** Check if a job has been applied to */
  function hasAppliedToJob(id) {
    return jobIds.has(id);
  }

  /** Apply to a job. Make API call to backend and update set of job ids. */
  function applyToJob(id) {
    if (hasAppliedToJob(id)) return;
    JoblyApi.applyToJob(currentUser.username, id);
    setJobIds(new Set([...jobIds, id]));
  }

  /**show a loading message while getting initial data */
  if (isLoading) {
    return <h1>Loading &hellip;</h1>;
  }

  return (
    <>
      <UserContext.Provider
        value={{ currentUser, setCurrentUser, hasAppliedToJob, applyToJob }}
      >
        <div className='App'>
          <NavBar logout={logout} />
          <ReactRoutes login={login} signup={signup} />
        </div>
      </UserContext.Provider>
    </>
  );
}

export default App;
