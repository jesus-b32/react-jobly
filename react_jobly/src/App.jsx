import { useState, useEffect } from 'react';
// import './App.css';
import { Route, Routes } from 'react-router-dom';
import JoblyApi from '../api';
//components
import HomePage from './Homepage';
import CompanyList from './CompanyList';
import CompanyDetail from './CompanyDetail';
import JobList from './JobList';
import NewUserForm from './NewUserForm';
import EditUserForm from './EditUserForm';
import LoginUserForm from './LoginUserForm';
import NavBar from './NavBar';
import NotFound from './NotFoundPage';

function App() {
  const [isLoading, setIsLoading] = useState(true);

  //store the list of companies; updated on first render and when new item added
  const [companies, setCompanies] = useState([]);

  //store the list of companies; updated on first render and when new item added
  const [jobs, setJobs] = useState([]);

  useEffect(() => {
    async function getCompanies() {
      const companies = await JoblyApi.getCompanies();
      const jobs = await JoblyApi.getJobs();
      setCompanies(companies);
      setJobs(jobs);
      setIsLoading(false);
    }
    getCompanies();
  }, []);

  // useEffect(() => {
  //   async function getCompany(companyHandle) {
  //     const company = await JoblyApi.getCompany(companyHandle);
  //     console.log('company', company);
  //     setCompany(company);
  //     // setIsLoading(false);
  //   }
  //   getCompany(companyHandle);
  //   // console.log('companies: ', companies);
  // }, [companyHandle]);

  if (isLoading) {
    return <h1>Loading &hellip;</h1>;
  }

  return (
    <>
      <NavBar />
      <Routes>
        <Route path='/' element={<HomePage />} />
        <Route
          path='/companies'
          element={<CompanyList companies={companies} />}
        />
        <Route path='/companies/:handle' element={<CompanyDetail />} />
        <Route path='/jobs' element={<JobList jobs={jobs} />} />
        <Route path='/login' element={<LoginUserForm />} />
        <Route path='/signup' element={<NewUserForm />} />
        <Route path='/profile' element={<EditUserForm />} />
        <Route path='/*' element={<NotFound />} />
      </Routes>
    </>
  );
}

export default App;
