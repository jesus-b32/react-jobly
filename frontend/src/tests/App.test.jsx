import { it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import App from '../App';
import { MemoryRouter } from 'react-router-dom';
import UserContext from '../auth/UserContext';

it('renders the App component', () => {
  render(
    <MemoryRouter>
      <App />
    </MemoryRouter>
  );
});

it('it matches snapshot', () => {
  const app = render(
    <MemoryRouter>
      <App />
    </MemoryRouter>
  );
  expect(app).toMatchSnapshot();
});

it('it displays Jobly title and slogan', async () => {
  render(
    <MemoryRouter>
      <App />
    </MemoryRouter>
  );
  expect(screen.getByText('Jobly')).toBeInTheDocument();
  expect(
    screen.getByText('All the jobs in one, convenient place.')
  ).toBeInTheDocument();
});

/**
 * A custom render to setup providers. Extends regular
 * render options with `providerProps` to allow injecting
 * different scenarios to test with.
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 */
const customRender = (ui, { providerProps, ...renderOptions }) => {
  return render(
    <UserContext.Provider {...providerProps}>
      <MemoryRouter>{ui}</MemoryRouter>
    </UserContext.Provider>,
    renderOptions
  );
};

/**
 * To test a component that provides a context value, render a matching
 * consumer as the child
 */
// it('NameConsumer shows value from provider', () => {
//   const currentUser = {
//     firstName: 'John',
//   };
//   customRender(<App />, { currentUser });
//   expect(screen.getByText('Welcome Back, John!')).toBeInTheDocument();
// });

/**
 * To test a component that provides a context value, render a matching
 * consumer as the child
 */
// it('NameProvider composes full name from first, last', () => {
//   const providerProps = {
//     first: 'Boba',
//     last: 'Fett',
//   }
//   customRender(
//     <UserContext.Consumer>
//       {(value) => <span>Received: {value}</span>}
//     </UserContext.Consumer>,
//     { providerProps }
//   );
//   expect(screen.getByText(/^Received:/).textContent).toBe('Received: Boba Fett')
// })
