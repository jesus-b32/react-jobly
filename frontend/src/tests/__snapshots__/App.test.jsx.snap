// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mb-4"
      >
        <nav
          class="navbar navbar-expand-md navbar-dark bg-dark"
        >
          <div
            class="container-fluid"
          >
            <a
              class="navbar-brand"
              href="/"
            >
              React-Jobly
            </a>
            <button
              aria-label="Toggle navigation"
              class="navbar-toggler"
              type="button"
            >
              <span
                class="navbar-toggler-icon"
              />
            </button>
            <div
              class="collapse navbar-collapse"
            >
              <ul
                class="ms-auto nav"
              >
                <li
                  class="ms-4 nav-item"
                >
                  <a
                    class="text-decoration-none text-white bg-dark"
                    href="/login"
                  >
                    Login
                  </a>
                </li>
                <li
                  class="mx-4 nav-item"
                >
                  <a
                    class="text-decoration-none text-white bg-dark"
                    href="/signup"
                  >
                    Sign Up
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </nav>
      </div>
      <div>
        <div
          class="d-flex flex-column home container-fluid"
        >
          <h1>
            <b>
              Jobly
            </b>
          </h1>
          <h4>
            All the jobs in one, convenient place.
          </h4>
          <div>
            <a
              class="btn btn-primary font-weight-bold me-3"
              href="/login"
            >
              Log in
            </a>
            <a
              class="btn btn-primary font-weight-bold"
              href="/signup"
            >
              Sign up
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mb-4"
    >
      <nav
        class="navbar navbar-expand-md navbar-dark bg-dark"
      >
        <div
          class="container-fluid"
        >
          <a
            class="navbar-brand"
            href="/"
          >
            React-Jobly
          </a>
          <button
            aria-label="Toggle navigation"
            class="navbar-toggler"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
            />
          </button>
          <div
            class="collapse navbar-collapse"
          >
            <ul
              class="ms-auto nav"
            >
              <li
                class="ms-4 nav-item"
              >
                <a
                  class="text-decoration-none text-white bg-dark"
                  href="/login"
                >
                  Login
                </a>
              </li>
              <li
                class="mx-4 nav-item"
              >
                <a
                  class="text-decoration-none text-white bg-dark"
                  href="/signup"
                >
                  Sign Up
                </a>
              </li>
            </ul>
          </div>
        </div>
      </nav>
    </div>
    <div>
      <div
        class="d-flex flex-column home container-fluid"
      >
        <h1>
          <b>
            Jobly
          </b>
        </h1>
        <h4>
          All the jobs in one, convenient place.
        </h4>
        <div>
          <a
            class="btn btn-primary font-weight-bold me-3"
            href="/login"
          >
            Log in
          </a>
          <a
            class="btn btn-primary font-weight-bold"
            href="/signup"
          >
            Sign up
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
